atc/worker/client.go

```diff
@@ -369,3 +493,148 @@ type processStatus struct {
 	processStatus int
 	processErr    error
 }
+
+func (client *client) RunPutStep(
+	ctx context.Context,
+	logger lager.Logger,
+	owner db.ContainerOwner,
+	containerSpec ContainerSpec,
+	workerSpec WorkerSpec,
+	strategy ContainerPlacementStrategy,
+	metadata db.ContainerMetadata,
+	imageFetcherSpec ImageFetcherSpec,
+	spec runtime.ProcessSpec,
+	eventDelegate runtime.StartingEventDelegate,
+	resource resource.Resource,
+) PutResult {
+
+	vr := runtime.VersionResult{}
+	err := client.wireInputsAndCaches(logger, &containerSpec)
+	if err != nil {
+
+		// TODO (runtime) Does it make sense to return -1 here?
+		return PutResult{Status: -1, VersionResult: runtime.VersionResult{}, Err: err}
+	}
+
+	chosenWorker, err := client.pool.FindOrChooseWorkerForContainer(
+		ctx,
+		logger,
+		owner,
+		containerSpec,
+		workerSpec,
+		strategy,
+	)
+	if err != nil {
+		return PutResult{Status: -1, VersionResult: runtime.VersionResult{}, Err: err}
+	}
+
+	container, err := chosenWorker.FindOrCreateContainer(
+		ctx,
+		logger,
+		imageFetcherSpec.Delegate,
+		owner,
+		metadata,
+		containerSpec,
+		imageFetcherSpec.ResourceTypes,
+	)
+	if err != nil {
+		return PutResult{Status: -1, VersionResult: runtime.VersionResult{}, Err: err}
+	}
+
+	// container already exited
+	exitStatusProp, err := container.Property(taskExitStatusPropertyName)
+	if err == nil {
+		logger.Info("already-exited", lager.Data{"status": exitStatusProp})
+
+		status, err := strconv.Atoi(exitStatusProp)
+		if err != nil {
+			// TODO (runtime) This is more confusing. We should differentiate failure vs. error very explicitly
+			return PutResult{-1, runtime.VersionResult{}, err}
+		}
+		return PutResult{Status: status, VersionResult: runtime.VersionResult{}, Err: nil}
+	}
+
+	eventDelegate.Starting(logger)
+
+	vr, err = resource.Put(ctx, spec, container)
+	if err != nil {
+		if failErr, ok := err.(runtime.ErrResourceScriptFailed); ok {
+			return PutResult{failErr.ExitStatus, runtime.VersionResult{}, failErr}
+		} else {
```

Comment:
IIRC, we tend to drop the `else` in these cases where there's an early return on an `if`.